harness/determined/searcher/search_runner.py

```diff
@@ -0,0 +1,243 @@
+import json
+import logging
+import os
+import time
+import uuid
+from pathlib import Path
+from typing import Any, Dict, Optional
+
+from determined.common.api import bindings
+from determined.common.api.bindings import v1SearcherEvent, v1TrialExitedEarlyExitedReason
+from determined.experimental import client
+from determined.searcher.search_method import ExitedReason, Progress, SearchMethod
+
+EXPERIMENT_ID_FILE = "experiment_id"
+STATE_FILE = "state"
+
+
+class SearchRunner:
+    def __init__(
+        self,
+        search_method: SearchMethod,
+    ) -> None:
+        self.search_method = search_method
+
+    def run_experiment(self, experiment_id: int) -> None:
+        assert client._determined is not None
+        session = client._determined._session
+
+        experiment_is_active = True
+
+        try:
+            while experiment_is_active:
+                time.sleep(5)
+                events = bindings.get_GetSearcherEvents(session, experimentId=experiment_id)
+                if events.searcherEvents is None:
+                    continue
+                logging.info(
+                    json.dumps(
+                        [SearchRunner._searcher_event_as_dict(e) for e in events.searcherEvents]
+                    )
+                )
+                # the first event is an event we have already processed and told master about it
+                # however, we may not have saved the state after that event if we crashed
+                # after POSTing operations but before saving state
+                last_event_id = self.search_method.searcher_state.last_event_id
+                first_event = True
+                for event in events.searcherEvents:
+                    assert event.id is not None
```

Comment:
is this an issue with the rest API? If all events should have an ID, why is it a nullable type at all?

Reply:
We will try to fix this.